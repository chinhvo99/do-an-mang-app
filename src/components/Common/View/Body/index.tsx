import React, { PureComponent } from 'react';
import {
  ScrollView, StyleSheet, SafeAreaView,
} from 'react-native';
import { withTheme, ThemeProps } from 'react-native-elements';
import QuickView, { QuickViewProps } from '../QuickView';

export interface BodyProps extends QuickViewProps {
  scroll?: boolean;
  theme?: any;
  secondary?: boolean;
  primary?: boolean;
  fullWidth?: boolean;
  fullHeight?: boolean;
  fullView?: boolean;
}

class Body extends PureComponent<BodyProps> {
  static defaultProps = {
    scroll: false,
    primary: true,
  };

  render() {
    const {
      scroll,
      backgroundColor: backgroundColorProp,
      paddingHorizontal: paddingHorizontalProp,
      secondary,
      fullView,
      fullWidth,
      fullHeight,
      theme,
      style: styleProp,
      children,
      ...otherProps
    } = this.props;

    const backgroundColor = backgroundColorProp
    || (secondary ? theme.colors.bgColorSecondary : theme.colors.bgColor);
    const paddingHorizontal = (fullWidth || fullView) ? 0
      : (paddingHorizontalProp || theme.bodyPaddingHorizontal);

    const style = StyleSheet.flatten([
      { backgroundColor, paddingHorizontal },
      styleProp,
    ]);

    if (fullHeight || fullView) {
      if (scroll) {
        return (
          <ScrollView testID="BodyScrollView" style={{ backgroundColor, flex: 1 }}>
            <QuickView {...otherProps} style={style}>
              {children}
            </QuickView>
          </ScrollView>
        );
      }
      return (
        <QuickView testID="FullHeightBodyView" {...otherProps} style={style}>
          {children}
        </QuickView>
      );
    }

    if (scroll) {
      return (
        <SafeAreaView testID="SafeAreaBodyScrollView" style={{ flex: 1 }}>
          <ScrollView style={{ backgroundColor }}>
            <QuickView {...otherProps} style={style}>
              {children}
            </QuickView>
          </ScrollView>
        </SafeAreaView>
      );
    }
    return (
      <SafeAreaView testID="SafeAreaBodyView" style={{ flex: 1 }}>
        <QuickView {...otherProps} style={style}>
          {children}
        </QuickView>
      </SafeAreaView>
    );
  }
}

export default withTheme(Body as React.ComponentType<BodyProps & ThemeProps<any>>);
