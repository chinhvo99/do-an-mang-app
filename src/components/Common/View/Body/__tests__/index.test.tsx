import React from 'react';
import {
  basicTest, EnumComponent, getComponent,
} from '@utils/testHelper';
import { lightTheme } from '@themes';
import Body from '..';

const target = <Body />;
const type = EnumComponent.HOC;

describe('Body Component', () => {
  basicTest(target, type, 2);

  it('should apply all custom props', () => {
    /**
     * scroll
     */
    const props = getComponent(target, { scroll: true }, type).props();
    expect(props.testID).toBe('SafeAreaBodyScrollView');

    const props2 = getComponent(target, {
      fullView: true, fullHeight: true, scroll: true,
    }, type).props();
    expect(props2.testID).toBe('BodyScrollView');

    /**
     * primary, secondary
     */
    const props3 = getComponent(target, { primary: true }, type).props();
    expect(props3.children.props.style.backgroundColor).toBe(lightTheme.colors.bgColor);

    const props4 = getComponent(target, { secondary: true }, type).props();
    expect(props4.children.props.style.backgroundColor).toBe(lightTheme.colors.bgColorSecondary);

    /**
     * fullWidth, fullHeight, fullView
     */
    const props5 = getComponent(target, { fullWidth: true }, type).props();
    expect(props5.children.props.style.paddingHorizontal).toBe(0);
    expect(props5.testID).toBe('SafeAreaBodyView');

    const props6 = getComponent(target, { fullHeight: true }, type).props();
    expect(props6.testID).toBe('FullHeightBodyView');
    expect(props6.style.paddingHorizontal).toBe(lightTheme.bodyPaddingHorizontal);

    const props7 = getComponent(target, { fullView: true }, type).props();
    expect(props7.testID).toBe('FullHeightBodyView');
    expect(props7.style.paddingHorizontal).toBe(0);
  });
});
