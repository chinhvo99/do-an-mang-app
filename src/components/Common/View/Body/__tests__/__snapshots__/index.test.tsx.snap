// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Body Component should render without issues 1`] = `
<ForwardRef(SafeAreaView)
  style={
    Object {
      "flex": 1,
    }
  }
  testID="SafeAreaBodyView"
>
  <QuickView
    column={true}
    primary={true}
    style={
      Object {
        "backgroundColor": "#ffffff",
        "paddingHorizontal": 18,
      }
    }
  />
</ForwardRef(SafeAreaView)>
`;
